rfc-087-use-libraries-where-possible.md

```diff
@@ -0,0 +1,128 @@
+and increase the potential for error.
+
+## Proposal
+
+When we need to add new generic functionality to our code, we have a
+choice of whether to write our own implementation, or use an
+off-the-shelf open source library.
+
+By default, we SHOULD aim to find and reuse existing open source
+libraries that fulfil the requirement rather than writing our own.
+
+Using existing code is preferable in order to get more proven
+code, a lower maintenance burden, and additional useful features we
+might not have time to implement ourselves.
+
+Some examples might include:
+
+- Date handling and formatting
+- Language inflection (e.g. pluralisation)
+- Queueing and background job running
+- Connecting to remote APIs
+
+The need for this will differ across languages and frameworks. For
+instance, Rails provides many capabilities that separate libraries
+would be needed for when working in Node.js. Conversely, frontend
+code may wish to avoid dependencies to keep file sizes low.
+
+Acquiring dependencies is not free, however. Developers MUST check
+them for licence compatibility, and SHOULD assess them for quality
+and reliability, so that we are sure they will be a net positive
+addition to our projects.
+
+### License Compatibility
+
+Our code is MIT licensed by default. All dependencies MUST have
+compatible licensing. In particular this means looking out for
+accidental aquisition of a dependency on GPL software (or other
+[copyleft](https://en.wikipedia.org/wiki/Copyleft) licenses that put
+license conditions on derivatives).
+
+You can use tools to check license compatibility in your entire
+dependency tree:
+
+- Ruby: `bundle licenses | sed 's/.*: //g' | sort | uniq`
+- Node.js: `npx license-checker --summary`
+  ([setup instructions](https://www.npmjs.com/package/license-checker))
+- Alternatives:
+  - [LicenseFinder](https://github.com/pivotal/LicenseFinder)
+  - [delice](https://github.com/cutenode/delice)
+  - [licensed](https://github.com/github/licensed) and
+    [bundler-licensed](https://github.com/sergey-alekseev/bundler-licensed)
+
+In general, [permissive licenses](https://en.wikipedia.org/wiki/Permissive_software_license)
+are considered to be MIT compatible, in particular:
+
+- [Apache](https://opensource.org/licenses/Apache-2.0)
+- BSD variants:
+  - [3-Clause](https://opensource.org/licenses/BSD-3-Clause)
+  - [2-Clause](https://opensource.org/licenses/BSD-2-Clause)
+  - [0BSD](https://opensource.org/licenses/0BSD)
+- [ISC](https://opensource.org/licenses/ISC)
+- [MPL](https://opensource.org/licenses/MPL-2.0)
+- [Unlicense](https://opensource.org/licenses/unlicense)
+- [Public Domain](https://en.wikipedia.org/wiki/Public-domain-equivalent_license)
+
+### Quality and Reliability
+
+We SHOULD only use dependencies that are maintained, trustworthy,
+and of a decent quality.
+
+It is nigh-impossible for us to inspect the source code of all our
+dependencies, however, so developer MAY use proxy measures to judge
```

Comment:
Is it worth explicitly allowing proxy measures like "Rails uses it"? If we're already depending on Rails, that's a non issue, but if we're doing a non-Rails thing in Ruby for some reason, the fact that it is a dependency for Rails is probably good enough.

Reply:
Good idea, I'll add that explicitly.

Reply:
Done (a bit further down)